File: sdk/cpp/examples/basic/basic.cpp
/**
 * KCM C++ SDK Basic Example
 *
 * Demonstrates RAII, exception safety, and modern C++ patterns.
 * Compile: g++ -std=c++17 -Wall -Wextra -O2 -I../../include -o basic basic.cpp -lkcm
 */

#include "kcm.hpp"
#include <iostream>
#include <cassert>

int main() {
    std::cout << "=== KCM C++ SDK Basic Example ===" << std::endl << std::endl;

    try {
        kcm::Database db;

        db.insert({1, 0, 2, 0.95, 0, 0, 0, 0, 0, 0});
        db.insert({2, 1, 3, 0.90, 0, 0, 0, 0, 0, 0});
        db.insert({3, 2, 4, 0.85, 0, 0, 0, 0, 0, 0});
        std::cout << "Inserted 3 facts" << std::endl;

        auto results = db.queryAll();
        std::cout << "Query returned " << results.size() << " facts" << std::endl;

        {
            auto txn = db.begin_transaction();
            db.insert({4, 3, 5, 0.80, 0, 0, 0, 0, 0, 0});
            txn.commit();
            std::cout << "Committed transaction" << std::endl;
        }

        db.remove(0);
        std::cout << "Deleted row 0" << std::endl;

        std::cout << "Fact count: " << db.fact_count() << std::endl;
        std::cout << "Active: " << db.active_count() << std::endl;

        try {
            kcm::Database::verify("/nonexistent.kcm");
        } catch (const kcm::Error& e) {
            std::cout << "Expected error: " << e.what() << std::endl;
        }

    } catch (const std::exception& e) {
        std::cerr << "Unexpected error: " << e.what() << std::endl;
        return 1;
    }

    std::cout << std::endl << "All C++ SDK examples passed!" << std::endl;
    return 0;
}

File: sdk/rust/examples/basic.rs
use kcm_sdk::{Database, Fact};

fn main() -> Result<(), Box<dyn std::error::Error>> {
    let db = Database::new()?;
    println!("Created database");

    let fact1 = Fact::new(1, 2, 3, 0.95)?
        .with_evidence(1)
        .with_context(1)
        .with_priority(0)
        .with_owner(1);
    let row1 = db.insert(&fact1)?;
    println!("Inserted fact at row {}", row1);

    let fact2 = Fact::new(4, 5, 6, 0.85)?
        .with_evidence(2)
        .with_context(1)
        .with_priority(1)
        .with_owner(2);
    let row2 = db.insert(&fact2)?;
    println!("Inserted fact at row {}", row2);

    let fact3 = Fact::new(1, 5, 9, 0.75)?
        .with_evidence(3)
        .with_context(2)
        .with_priority(0)
        .with_owner(3);
    db.insert(&fact3)?;

    println!("Total facts: {}", db.fact_count());
    println!("Active facts: {}", db.active_fact_count());

    let results = db.query("all")?;
    println!("Query returned {} results:", results.count());
    for fact in results {
        println!(
            "  S={} P={} O={} conf={:.2}",
            fact.subject, fact.predicate, fact.object, fact.confidence
        );
    }

    let updated = Fact::new(10, 20, 30, 0.99)?;
    db.update(row1, &updated)?;
    println!("Updated fact at row {}", row1);

    let retrieved = db.get_fact(row1)?.expect("fact not found");
    println!(
        "Retrieved: S={} P={} O={} conf={:.2}",
        retrieved.subject, retrieved.predicate, retrieved.object, retrieved.confidence
    );

    db.delete(row2)?;
    println!("Deleted fact at row {}", row2);
    println!("Active facts after delete: {}", db.active_fact_count());

    let mut txn = db.begin_transaction()?;
    let txn_fact = Fact::new(100, 200, 300, 0.6)?;
    txn.insert(&txn_fact)?;
    println!("Transaction changes: {}", txn.change_count());
    txn.commit()?;
    println!("Transaction committed");
    println!("Final fact count: {}", db.fact_count());

    let dir = tempfile::tempdir()?;
    let path = dir.path().join("example.kcm");
    let path_str = path.to_str().expect("invalid path");
    db.save(path_str)?;
    println!("Saved database to {}", path_str);

    Database::verify(path_str)?;
    println!("Database verification passed");

    let loaded = Database::load(path_str)?;
    println!("Loaded database with {} facts", loaded.fact_count());

    db.close();
    println!("Done!");

    Ok(())
}
